test/README.md

```diff
@@ -12,9 +14,219 @@ To run all unit tests:
+the caller to identify what state is being polled for._
+
+### Check Knative Serving resources
+
+_WARNING: this code also exists in
+[`knative/serving`](https://github.com/knative/serving/blob/master/test/adding_tests.md#make-requests-against-deployed-services)._
+
+After creating Knative Serving resources or making changes to them, you will need to wait for the system
+to realize those changes. You can use the Knative Serving CRD check and polling methods to check the
+resources are either in or reach the desired state.
+
+The `WaitFor*` functions use the kubernetes [`wait` package](https://godoc.org/k8s.io/apimachinery/pkg/util/wait).
+To poll they use [`PollImmediate`](https://godoc.org/k8s.io/apimachinery/pkg/util/wait#PollImmediate)
+and the return values of the function you provide behave the same as
+[`ConditionFunc`](https://godoc.org/k8s.io/apimachinery/pkg/util/wait#ConditionFunc):
+a `bool` to indicate if the function should stop or continue polling, and an `error` to indicate if
+there has been an error.
+
+For example, you can poll a `Configuration` object to find the name of the `Revision` that was created
+for it:
+
+```go
+var revisionName string
+err := test.WaitForConfigurationState(clients.ServingClient, configName, func(c *v1alpha1.Configuration) (bool, error) {
+    if c.Status.LatestCreatedRevisionName != "" {
+        revisionName = c.Status.LatestCreatedRevisionName
+        return true, nil
+    }
+    return false, nil
+}, "ConfigurationUpdatedWithRevision")
+```
+
+_[Metrics will be emitted](#emit-metrics) for these `Wait` method tracking how long test poll for._
+
+_See [kube_checks.go](./kube_checks.go)._
+
+### Ensure test cleanup
+
+To ensure your test is cleaned up, you should defer cleanup to execute after your
+test completes and also ensure the cleanup occurs if the test is interrupted:
+
+```go
+defer tearDown(clients)
+test.CleanupOnInterrupt(func() { tearDown(clients) })
+```
+
+_See [cleanup.go](./cleanup.go)._
+
+## Flags
+
+Importing [the test library](#test-library) adds flags that are useful for end to end tests that
+need to run against a cluster.
+
+Tests importing [`github.com/knative/pkg/test`](#test-library) recognize these flags:
+
+* [`--kubeconfig`](#specifying-kubeconfig)
+* [`--cluster`](#specifying-cluster)
+* [`--namespace`](#specifying-namespace)
+* [`--logverbose`](#output-verbose-logs)
+* [`--emitmetrics`](#metrics-flag)
+
+### Specifying kubeconfig
+
+By default the tests will use the [kubeconfig
+file](https://kubernetes.io/docs/concepts/configuration/organize-cluster-access-kubeconfig/)
+at `~/.kube/config`. If there is an error getting the current user, it will use `kubeconfig` instead as the default value.
+You can specify a different config file with the argument `--kubeconfig`.
+
+To run the tests with a non-default kubeconfig file:
+
+```bash
+go test -v -tags=e2e -count=1 ./test/conformance --kubeconfig /my/path/kubeconfig
```

Comment:
The examples here are specific to serving. Either we should not have them, or explicitly state that this is an example from serving

Reply:
hmmm let me see what I can do, let me know if it's still not right